## D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?


## Steps

Import all the libraries

In [78]:
import pandas as pd
import numpy as np
import sklearn.metrics as metrics
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense

import warnings
warnings.simplefilter('ignore', FutureWarning)

## Reading the data

In [79]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

#### 1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [80]:
# Getting the features and targets
features = concrete_data.drop(columns='Strength', axis=1)
Y = concrete_data['Strength']

features = (features - features.mean()) / features.std()

n_cols = features.shape[1] # number of predictors

x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)


#### 2. Train the model on the training data using 50 epochs.

In [81]:
def regression_model():
    # create model
    model = Sequential()
    #One hidden layer of 10 nodes, and a ReLU activation function 
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model = regression_model()
model

#### 2. Train the model on the training data using 50 epochs.

In [82]:
model.fit(x_train, y_train, epochs=100, verbose=2)

Epoch 1/100
 - 1s - loss: 1488.6574
Epoch 2/100
 - 0s - loss: 1442.3837
Epoch 3/100
 - 0s - loss: 1374.2357
Epoch 4/100
 - 0s - loss: 1272.6569
Epoch 5/100
 - 0s - loss: 1130.5918
Epoch 6/100
 - 0s - loss: 955.3693
Epoch 7/100
 - 0s - loss: 757.1256
Epoch 8/100
 - 0s - loss: 563.9981
Epoch 9/100
 - 0s - loss: 409.0752
Epoch 10/100
 - 0s - loss: 308.3321
Epoch 11/100
 - 0s - loss: 255.8955
Epoch 12/100
 - 0s - loss: 232.5622
Epoch 13/100
 - 0s - loss: 215.1094
Epoch 14/100
 - 0s - loss: 203.8705
Epoch 15/100
 - 0s - loss: 195.1585
Epoch 16/100
 - 0s - loss: 188.6652
Epoch 17/100
 - 0s - loss: 183.0311
Epoch 18/100
 - 0s - loss: 178.4899
Epoch 19/100
 - 0s - loss: 173.9785
Epoch 20/100
 - 0s - loss: 170.2786
Epoch 21/100
 - 0s - loss: 167.1161
Epoch 22/100
 - 0s - loss: 164.0402
Epoch 23/100
 - 0s - loss: 161.5789
Epoch 24/100
 - 0s - loss: 159.0643
Epoch 25/100
 - 0s - loss: 156.4941
Epoch 26/100
 - 0s - loss: 154.4564
Epoch 27/100
 - 0s - loss: 152.6107
Epoch 28/100
 - 0s - loss: 150.8

#### 3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [83]:
predictions = model.predict(x_test)
MAE =  metrics.mean_absolute_error(y_test, predictions)
print(MAE)

7.975386483152322


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [84]:
MAEs = []
for n in range(50): 
    x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.3)
    model.fit(x_train, y_train, epochs=50, verbose=0)
    predictions = model.predict(x_test)
    MAEs.append(metrics.mean_absolute_error(y_test, predictions))
    
print(MAEs)    

[5.3126254665041435, 4.944411728374395, 4.780841490674559, 4.848509130014957, 4.985573995028499, 4.637996351865503, 4.819420279715825, 4.666688535375502, 4.399531115312793, 4.5437654816525646, 4.259345525352701, 4.149573304259661, 4.402319552381448, 4.0729418742232335, 3.84804248473405, 3.9179027430833733, 4.03941424132165, 4.230937591416936, 3.657551345053614, 3.6907517970804253, 3.848560000931946, 4.03372379488158, 3.944765863017357, 3.5519373343754745, 3.733020334614133, 3.754452292556516, 3.6346179798583, 3.5480513513435437, 3.2676643612315353, 3.4817009971211252, 3.847634152501918, 3.59821777590656, 3.587927431396682, 3.513110032066172, 3.4340079745196985, 3.247796299388107, 3.5059763659319834, 3.652175245470213, 3.478357217967703, 3.4429133451023537, 3.242886947088643, 3.272076534715671, 3.075155461752685, 3.7607078842360613, 3.6441966590140633, 3.5412102751438677, 3.3354231769598806, 3.580364490681868, 3.257196279852907, 3.197019080313278]


#### 5. Report the mean and the standard deviation of the mean squared errors.

In [85]:
np_array = np.array(MAEs)
print(f"Mean of mean squared errors:{np.mean(np_array)} ")
print(f"Standard deviation of mean squared errors:{np.std(np_array)} ")

Mean of mean squared errors:3.8844198194673534 
Standard deviation of mean squared errors:0.5454374320997948 


### How does the mean of the mean squared errors compare to that from Step B?


Step A:

- Mean of mean squared errors:5.57246862771433 
- Standard deviation of mean squared errors:0.7217945033478287 

Step B:
 - Mean of mean squared errors:5.718177185425865 
 - Standard deviation of mean squared errors:1.2205812457213703 
 
Step C
 - Mean of mean squared errors:4.931535241926989 
 - Standard deviation of mean squared errors:0.8790470154158412 
 
Step D
 - Mean of mean squared errors:3.8844198194673534 
 - Standard deviation of mean squared errors:0.5454374320997948 

Step D is a lot better if compared with Step B, with reduced mean and std deviation